clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -5276,9 +5277,31 @@ void Sema::InstantiateFunctionDefinition(SourceLocation PointOfInstantiation,
     RebuildTypeSourceInfoForDefaultSpecialMembers();
     SetDeclDefaulted(Function, PatternDecl->getLocation());
   } else {
+    NamedDecl *ND = Function;
+    DeclContext *DC = ND->getLexicalDeclContext();
+    std::optional<ArrayRef<TemplateArgument>> Innermost;
+    if (auto *Primary = Function->getPrimaryTemplate();
+        Primary &&
+        !isGenericLambdaCallOperatorOrStaticInvokerSpecialization(Function) &&
+        Function->getTemplateSpecializationKind() !=
+            TSK_ExplicitSpecialization) {
+      auto It = llvm::find_if(Primary->redecls(),
+                              [](const RedeclarableTemplateDecl *RTD) {
+                                return cast<FunctionTemplateDecl>(RTD)
+                                    ->isCompatibleWithDefinition();
+                              });
+      assert(It != Primary->redecls().end() &&
+             "Should't get here without a definition");
+      if (FunctionDecl *Def = cast<FunctionTemplateDecl>(*It)
+                                  ->getTemplatedDecl()
+                                  ->getDefinition())
+        DC = Def->getLexicalDeclContext();
+      else
+        DC = (*It)->getLexicalDeclContext();
```

Comment:
Can we get into a situation like `regression2` with friends?
I wonder if we can use `DC = (*It)->getDeclContext();` (not lexical!) unconditionally.

Reply:
We need the lexical context because we are interested in the template context of the declaration, ie what set of parameters it sits under, which is a lexical thing.

I don't follow what you mean this situation with friends, can you provide an example?
regression2 is a situation very specific to member function templates, but friends are not members.

Reply:
Sorry, nevermind. I understand what you are doing now.

However, I'm wondering if we are not reinventing `getTemplateInstantiationPattern(true)` here, It looks strangely similar. Have you explored that?

Reply:
Well, there is commonality in that both are using the same helpers, but `getTemplateInstantiationPattern(true)` doesn't do what we need here, and this bit doesn't do what the other users of `getTemplateInstantiationPattern(true)` need. I am not sure it would be worth joining both implementations, adding an extra parameter to `getTemplateInstantiationPattern`.